src/state/opamSwitchState.ml

```diff
@@ -849,8 +849,9 @@ let universe st
     ?reinstall
     ~requested
     user_action =
+  let names = OpamPackage.names_of_packages requested in
   let requested_allpkgs =
-    OpamPackage.packages_of_names st.packages requested
+    OpamPackage.packages_of_names st.packages names
```

Comment:
This is patched in a reasonable and conservative way...

But this makes me wonder: it would seem consistent to use `requested` here directly instead of `requested_allpkgs`. This would potentially change the semantics at some point, but could be expected to lead to the same results...

A visible change I could imagine is when you change the installed version of a package, _e.g._ with `opam install foo.2` when `foo.1` is installed: in the current implementation (and behaviour preserved by the proposed patch), `foo.1` is a member of `requested_allpkgs`. This affects:
- the values of `with-test`, `with-doc`: this is harmless, the action on `foo.1` can only be a removal
- the triggering of pending reinstallation within the dependency cone of `foo.1`

The second point _could_ have visible effects in some cases: the idea is not to trigger pending recompilations (see [`opam reinstall --list-pending`](http://opam.ocaml.org/doc/man/opam-reinstall.html)) if you `opam install` something unrelated. If `foo.2` depends on the package needing reinstallation, opam definitely needs to do that reinstallation before compiling it ; but if it's only `foo.1`, that will just be removed, I guess it can be skipped.

All well considered, I think this can be simplified by just using `requested` everywhere instead of `requested_allpkgs` :)